pkg/kubernetes/composite_client.go

```diff
@@ -0,0 +1,352 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package kubernetes
+
+import (
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+)
+
+// CompositeClientSet wraps a pair of Kubernetes clients hooked up to a live api server.
+//
+// Prefer this client when the cluster CA is expected to change (ex.: secret rotation operations).
+type CompositeClientSet struct {
+	oldCAClient Client
+	newCAClient Client
+	timeout     time.Duration
+	backoff     wait.Backoff
+	retryFunc   func(err error) bool
+}
+
+// NewCompositeClient returns a KubernetesClient hooked up to the api server at the apiserverURL.
+func NewCompositeClient(oldCAClient, newCAClient Client, interval, timeout time.Duration) *CompositeClientSet {
+	return &CompositeClientSet{
+		oldCAClient: oldCAClient,
+		newCAClient: newCAClient,
+		timeout:     timeout,
+		backoff: wait.Backoff{
+			Steps:    int(int64(timeout/time.Millisecond) / int64(interval/time.Millisecond)),
+			Duration: interval,
+			Factor:   1.0,
+			Jitter:   0.0,
+		},
+		retryFunc: func(err error) bool { return true }, // Always retry. Inject if ever needed
+	}
+}
+
+// ListPods returns Pods based on the passed in list options.
+func (c *CompositeClientSet) ListPods(namespace string, opts metav1.ListOptions) (*v1.PodList, error) {
```

Comment:
So each invocation of `ListPods` makes two calls, one for the old client and one for the new client, to its configured apiserver endpoint?

Reply:
(and obv. those calls are in parallel)

Reply:
correct